File: backend/permissions.py
"""Role → policy definitions and lookup helpers.

This module is the single source of truth for what each role *is* and *can do*.
Every role maps to a nested "policy object" that bundles two kinds of thing:

  * access permissions — e.g. ``control_panel.view`` (can this role open the
    Control Panel), and
  * role attributes / metadata — e.g. ``label`` (display name) and
    ``school_info.required`` (does onboarding ask this role for grade/team year).

Keeping both in one object means a role has exactly one definition instead of
flags scattered across the frontend and backend. The structure is intentionally
a nested dict so values can be booleans, numbers, strings, or further nesting
without changing the shape of callers.

This module is deliberately pure: it has no FastAPI / request dependencies so it
stays trivially testable. The request-time enforcement dependency
(``require_permission``) lives in ``endpoints.auth`` where ``get_current_user``
is available, to avoid a circular import.
"""

from typing import Any

# Each role → its full policy object. `control_panel.view` is an access
# permission; `school_info.required` and `label` are role attributes.
ROLE_DEFINITIONS: dict[str, dict[str, Any]] = {
    "cad_member":           {"label": "CAD Member",           "school_info": {"required": True},  "attendance": {"view": True},  "control_panel": {"view": False}},
    "cad_lead":             {"label": "CAD Lead",             "school_info": {"required": True},  "attendance": {"view": True},  "control_panel": {"view": True, "meeting_agenda": True}},
    "manufacturing_member": {"label": "Manufacturing Member", "school_info": {"required": True},  "attendance": {"view": True},  "control_panel": {"view": False}},
    "manufacturing_lead":   {"label": "Manufacturing Lead",   "school_info": {"required": True},  "attendance": {"view": True},  "control_panel": {"view": True, "meeting_agenda": True}},
    "programming_member":   {"label": "Programming Member",   "school_info": {"required": True},  "attendance": {"view": True},  "control_panel": {"view": False}},
    "programming_lead":     {"label": "Programming Lead",     "school_info": {"required": True},  "attendance": {"view": True},  "control_panel": {"view": True, "meeting_agenda": True}},
    "scouting_member":      {"label": "Scouting Member",      "school_info": {"required": True},  "attendance": {"view": True},  "control_panel": {"view": True, "meeting_time": True, "meeting_agenda": True, "upcoming_event": True}},
    "scouting_lead":        {"label": "Scouting Lead",        "school_info": {"required": True},  "attendance": {"view": True},  "control_panel": {"view": True, "meeting_time": True, "meeting_agenda": True, "upcoming_event": True}},
    "publicity_member":     {"label": "Publicity Member",     "school_info": {"required": True},  "attendance": {"view": True},  "control_panel": {"view": False}},
    "publicity_lead":       {"label": "Publicity Lead",       "school_info": {"required": True},  "attendance": {"view": True},  "control_panel": {"view": True, "meeting_agenda": True}},
    "operations_member":    {"label": "Operations Member",    "school_info": {"required": True},  "attendance": {"view": True},  "control_panel": {"view": False}},
    "operations_lead":      {"label": "Operations Lead",      "school_info": {"required": True},  "attendance": {"view": True},  "control_panel": {"view": True, "meeting_agenda": True}},
    "outreach_member":      {"label": "Outreach Member",      "school_info": {"required": True},  "attendance": {"view": True},  "control_panel": {"view": False}},
    "outreach_lead":        {"label": "Outreach Lead",        "school_info": {"required": True},  "attendance": {"view": True},  "control_panel": {"view": True, "meeting_agenda": True}},
    "captain":              {"label": "Captain",              "school_info": {"required": True},  "attendance": {"view": True},  "control_panel": {"view": True, "meeting_time": True, "meeting_agenda": True, "upcoming_event": True}},
    "mentor":               {"label": "Mentor",               "school_info": {"required": False}, "attendance": {"view": False}, "control_panel": {"view": True, "meeting_time": True, "meeting_agenda": True, "upcoming_event": True}},
    "alumni":               {"label": "Alumni",               "school_info": {"required": False}, "attendance": {"view": False}, "control_panel": {"view": False}},
}


def get_permissions_for_role(role: str | None) -> dict[str, Any]:
    """Return the full policy object for a role.

    Users hold a single role (see ``users.role``), so this is a direct lookup
    with no merging. Returns an empty dict for an unknown or missing role, which
    makes every downstream ``can(...)`` check evaluate to ``False`` — i.e. an
    unrecognised role is treated as having no permissions.

    :param role: The role slug (e.g. ``"scouting_lead"``) or ``None``.
    :returns: The role's policy object, or ``{}`` if the role is unknown/``None``.
    """
    if not role:
        return {}
    return ROLE_DEFINITIONS.get(role, {})


def get_perm(perms: dict[str, Any] | None, path: str) -> Any:
    """Safely read a value from a policy object by dotted path.

    Walks a ``"a.b.c"`` path through the nested policy dict, returning ``None``
    the moment any segment is missing rather than raising ``KeyError`` /
    ``TypeError``. This is the only supported way to read nested policy values —
    callers must never index the dict directly, so a missing branch can never
    crash a request.

    :param perms: A policy object (e.g. from :func:`get_permissions_for_role`).
    :param path: Dotted key path, e.g. ``"control_panel.view"``.
    :returns: The value at ``path``, or ``None`` if any segment is absent.
    """
    node: Any = perms
    for key in path.split("."):
        if not isinstance(node, dict) or key not in node:
            return None
        node = node[key]
    return node


def can(perms: dict[str, Any] | None, path: str) -> bool:
    """Return whether a policy grants a boolean capability at ``path``.

    Thin truthiness wrapper over :func:`get_perm` for the common yes/no case,
    e.g. ``can(perms, "control_panel.view")``. A missing path is ``False``.

    :param perms: A policy object.
    :param path: Dotted capability path, e.g. ``"control_panel.view"``.
    :returns: ``True`` if the value at ``path`` is truthy, else ``False``.
    """
    return bool(get_perm(perms, path))


def role_catalog() -> list[dict[str, Any]]:
    """Return the public list of roles for the onboarding role picker.

    During onboarding the user is authenticated but has not chosen a role yet,
    so the frontend needs metadata for *every* role up front (which ``/auth/me``
    cannot provide, as it only knows the current user's role). This exposes just
    the fields the picker needs — value, display label, and whether the role
    requires school info — derived from :data:`ROLE_DEFINITIONS`, preserving its
    declaration order.

    :returns: A list of ``{"value", "label", "school_info_required"}`` dicts.
    """
    return [
        {
            "value": role,
            "label": defn.get("label", role),
            "school_info_required": bool(get_perm(defn, "school_info.required")),
        }
        for role, defn in ROLE_DEFINITIONS.items()
    ]


__all__ = [
    "ROLE_DEFINITIONS",
    "get_permissions_for_role",
    "get_perm",
    "can",
    "role_catalog",
]

File: frontend/src/lib/permissions.ts
/**
 * Client-side helpers for reading the role policy sent by the backend.
 *
 * The backend (permissions.py) owns the role → policy map and ships the current
 * user's resolved policy on `/auth/me` as `user.permissions`. The frontend never
 * decides *what* a role can do — it only reads the verdict from that policy to
 * gate UI (which is cosmetic; the backend still enforces access). Mirror of the
 * backend `get_perm` / `can` helpers so both sides read the nested policy the
 * same, null-safe way.
 */

/** A value stored in a policy object: a leaf or a nested policy. */
export type PermValue = boolean | number | string | string[] | PermPolicy

/** A role policy object: nested access permissions + role attributes. */
export interface PermPolicy {
    [key: string]: PermValue
}

/** One entry in the role catalog returned by `GET /roles`. Used by the onboarding
 *  picker and by the Members roster to place a row (`level`/`subteam`) when
 *  deciding whether the current user may moderate it. */
export interface RoleCatalogEntry {
    value: string
    label: string
    school_info_required: boolean
    level: string
    subteam: string | null
    requires_approval: boolean
}

/** The `can_moderate` spec carried on a role policy (see backend permissions.py).
 *  `scope: "all"` clears any target; `scope: "subteam"` clears targets whose
 *  `level` is in `target_levels` and whose subteam matches the actor's, plus any
 *  role in `extra_roles`. Absent on roles with no moderation authority. */
interface ModerateSpec {
    scope?: string
    target_levels?: string[]
    extra_roles?: string[]
}

/**
 * Safely read a value from a policy object by dotted path.
 *
 * Walks an `"a.b.c"` path through the nested policy, returning `undefined` the
 * moment any segment is missing rather than throwing. This is the only supported
 * way to read nested policy values — never index the object directly, so a
 * missing branch can't crash the UI.
 *
 * @param perms A policy object (e.g. `user.permissions`), or null/undefined.
 * @param path Dotted key path, e.g. `"control_panel.view"`.
 * @returns The value at `path`, or `undefined` if any segment is absent.
 */
export function getPerm(perms: PermPolicy | null | undefined, path: string): PermValue | undefined {
    let node: PermValue | null | undefined = perms
    for (const key of path.split(".")) {
        if (node == null || typeof node !== "object" || Array.isArray(node) || !(key in node)) {
            return undefined
        }
        node = node[key]
    }
    return node
}

/**
 * Return whether a policy grants a boolean capability at `path`.
 *
 * Truthiness wrapper over {@link getPerm} for the common yes/no UI check, e.g.
 * `can(user.permissions, "control_panel.view")`. A missing path is `false`.
 *
 * @param perms A policy object, or null/undefined.
 * @param path Dotted capability path, e.g. `"control_panel.view"`.
 * @returns `true` if the value at `path` is truthy, else `false`.
 */
export function can(perms: PermPolicy | null | undefined, path: string): boolean {
    return Boolean(getPerm(perms, path))
}

/**
 * Return whether the current user (with policy `myPolicy`) may moderate a member
 * holding `targetRole` — the same approve/ban scope the backend enforces in
 * `can_role_moderate`. This is cosmetic (it only enables/disables roster buttons);
 * the endpoints re-check it. `catalog` supplies the target role's `level`/`subteam`.
 *
 * @param myPolicy The acting user's resolved policy (`user.permissions`).
 * @param targetRole The role slug of the member being acted on.
 * @param catalog The role catalog from `GET /roles`.
 * @returns `true` if the action is within the user's scope, else `false`.
 */
export function canModerate(
    myPolicy: PermPolicy | null | undefined,
    targetRole: string | null | undefined,
    catalog: RoleCatalogEntry[],
): boolean {
    const raw = getPerm(myPolicy, "can_moderate")
    if (raw == null || typeof raw !== "object" || Array.isArray(raw)) return false
    const spec = raw as unknown as ModerateSpec
    if (spec.scope === "all") return true
    if (spec.scope === "subteam") {
        if (targetRole != null && spec.extra_roles?.includes(targetRole)) return true
        const target = catalog.find(r => r.value === targetRole)
        if (!target || target.subteam == null) return false
        const mySubteam = getPerm(myPolicy, "subteam")
        return (spec.target_levels?.includes(target.level) ?? false) && target.subteam === mySubteam
    }
    return false
}
